dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,1233 @@
+      return fill_value;
+    }
+    return in[dot(coords, sample_shape.in_strides)];
+  }
+
+  In fill_value;
+};
+
+template <typename InLoader>
+struct InLoaderProvider {
+  DALI_DEVICE DALI_FORCEINLINE InLoader operator()(int sample_idx) {
+    return {};
+  }
+};
+
+template <typename In, int axes>
+struct InLoaderProvider<InLoaderPad<In, axes>> {
+  DALI_DEVICE DALI_FORCEINLINE InLoaderPad<In, axes> operator()(int sample_idx) {
+    if (fill_values == nullptr) {
+      return {0};
+    }
+    return {fill_values[sample_idx][0]};
+  }
+  const In** fill_values;
+};
+
+/** @} */  // end of InputLoader
+
+
+/** @defgroup OutputShapeProvider OutputShapeProvider is introduced to reduce registers pressure
+ * when it is known that input and output shapes are equal.
+ * @{
+ */
+
+template <typename SampleDescT>
+struct OutShapeProviderSame {
+  DALI_DEVICE DALI_FORCEINLINE OutShapeDesc<SampleDescT::axes> operator()(SampleDescT sample_desc) {
+    auto in_shape = sample_desc.in_shape;
+    return {in_shape.frame_stride, in_shape.in_strides, in_shape.in_extents};
+  }
+};
+
+template <typename SampleDescT>
+struct OutShapeProviderROI {
+  DALI_DEVICE DALI_FORCEINLINE OutShapeDesc<SampleDescT::axes> operator()(SampleDescT sample_desc) {
+    return sample_desc.out_shape;
+  }
+};
+
+/** @} */  // end of OutputShapeProvider
+
+
+/** @defgroup InputConv The specializations provide ``compute`` method that computes convolution
+ * for the patch of logical block size (logical_block_extents) and stores it in the provided
+ * ``acc``.
+ * @{
+ */
+
+/**
+ * @brief First loads the input patch necessary to compute the output of logical block size
+ * (logical_block_extents) into shared memory (including filter's halo/apron), then computes the
+ * convolution.
+ */
+template <int lane_axis_, typename SampleDescT_, typename Inloader_, typename BlockSetupT_>
+class ShmInputConv {
+ public:
+  using SampleDescT = SampleDescT_;
+  using Inloader = Inloader_;
+  using BlockSetupT = BlockSetupT_;
+  using StaticConfigT = typename BlockSetupT::StaticConfigT;
+  using In = typename SampleDescT::In;
+  using Acc = typename SampleDescT::Acc;
+  static constexpr int lane_axis = lane_axis_;
+
+  DALI_DEVICE DALI_FORCEINLINE ShmInputConv(SampleDescT sample_desc, Inloader in_loader,
+                                            BlockSetupT block_setup, In* in_workspace,
+                                            int* precomputed_idx)
+      : sample_desc_{sample_desc},
+        in_loader_{in_loader},
+        block_setup_{block_setup},
+        in_workspace_{in_workspace},
+        precomputed_idx_{precomputed_idx} {}
+
+  DALI_DEVICE DALI_FORCEINLINE void compute(Acc* acc, const In* __restrict__ in,
+                                            ivec<SampleDescT::axes> start) const {
+    auto anchored_start = start - sample_desc_.in_shape.anchor_shift;
+    __syncthreads();
+    precompute_indices(in, anchored_start);
```

Comment:
`in` seems unused?

Reply:
Spot-on, I guess it is an artifact of splitting the precomuting and loading into two functions.